cps/kobo.py

```diff
@@ -0,0 +1,508 @@
+def to_epoch_timestamp(datetime_object):
+    return (datetime_object - datetime(1970, 1, 1)).total_seconds()
+
+
+def get_store_url_for_current_request():
+    # Programmatically modify the current url to point to the official Kobo store
+    base, sep, request_path_with_auth_token = request.full_path.rpartition("/kobo/")
+    auth_token, sep, request_path = request_path_with_auth_token.rstrip("?").partition(
+        "/"
+    )
+    return KOBO_STOREAPI_URL + "/" + request_path
+
+
+CONNECTION_SPECIFIC_HEADERS = [
+    "connection",
+    "content-encoding",
+    "content-length",
+    "transfer-encoding",
+]
+
+
+def redirect_or_proxy_request():
+    if request.method == "GET":
+        return redirect(get_store_url_for_current_request(), 307)
+    else:
+        # The Kobo device turns other request types into GET requests on redirects, so we instead proxy to the Kobo store ourselves.
+        outgoing_headers = Headers(request.headers)
+        outgoing_headers.remove("Host")
+        store_response = requests.request(
+            method=request.method,
+            url=get_store_url_for_current_request(),
+            headers=outgoing_headers,
+            data=request.get_data(),
+            allow_redirects=False,
+        )
+
+        response_headers = store_response.headers
+        for header_key in CONNECTION_SPECIFIC_HEADERS:
+            response_headers.pop(header_key, default=None)
+
+        return make_response(
+            store_response.content, store_response.status_code, response_headers.items()
+        )
+
+
+class SyncToken:
+    """ The SyncToken is used to persist state accross requests.
+    When serialized over the response headers, the Kobo device will propagate the token onto following requests to the service.
+    As an example use-case, the SyncToken is used to detect books that have been added to the library since the last time the device synced to the server.
+
+    Attributes:
+        books_last_created: Datetime representing the newest book that the device knows about.
+        books_last_modified: Datetime representing the last modified book that the device knows about.
+    """
+
+    SYNC_TOKEN_HEADER = "x-kobo-synctoken"
+    VERSION = "1-0-0"
+    MIN_VERSION = "1-0-0"
+
+    token_schema = {
+        "type": "object",
+        "properties": {"version": {"type": "string"}, "data": {"type": "object"},},
+    }
+    # This Schema doesn't contain enough information to detect and propagate book deletions from Calibre to the device.
+    # A potential solution might be to keep a list of all known book uuids in the token, and look for any missing from the db.
+    data_schema_v1 = {
+        "type": "object",
+        "properties": {
+            "raw_kobo_store_token": {"type": "string"},
+            "books_last_modified": {"type": "string"},
+            "books_last_created": {"type": "string"},
+        },
+    }
+
+    def __init__(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Possible hardcoded password: ''](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

Reply:
I'm not sure we should worry too much about this sort of false-positive 'harcoded password' check.

Reply:
It's no problem, maybe I should deactivate the check in codacy.